tortoise/verifying_tortoise.go

```diff
@@ -194,428 +309,936 @@ func (t *turtle) checkBlockAndGetInputVector(
+		return nil, fmt.Errorf("%s (%v)", errstrTooManyExceptions, explen)
 	}
 
 	return []map[types.BlockID]struct{}{againstDiff, forDiff, neutralDiff}, nil
 }
 
-func (t *turtle) BlockWeight(voting, voted types.BlockID) int {
-	return 1
+// voteWeight returns the weight to assign to one block's vote for another.
+// Note: weight depends on more than just the weight of the voting block. It also depends on contextual factors such as
+// whether or not the block's ATX was received on time, and on how old the layer is.
+// TODO: for now it's probably sufficient to adjust weight based on whether the ATX was received on time, or late, for
+//   the current epoch. See https://github.com/spacemeshos/go-spacemesh/issues/2540.
+// TODO: does blockVotedOn ever matter here? can we remove it?
+func (t *turtle) voteWeight(ctx context.Context, votingBlock *types.Block, blockVotedOn types.BlockID) (uint64, error) {
+	logger := t.logger.WithContext(ctx)
+
+	atxHeader, err := t.atxdb.GetAtxHeader(votingBlock.ATXID)
+	if err != nil {
+		return 0, err
+	}
+	atxTimestamp, err := t.atxdb.GetAtxTimestamp(votingBlock.ATXID)
+	if err != nil {
+		return 0, err
+	}
+	atxEpoch := atxHeader.PubLayerID.GetEpoch()
+	nextEpochStart := t.clock.LayerToTime((atxEpoch + 1).FirstLayer())
+
+	// check if the ATX was received on time
+	// TODO LANE: add an exception for sync, when we expect everything to be received late
+	if atxTimestamp.Before(nextEpochStart) {
+		blockWeight := atxHeader.GetWeight()
+		logger.With().Debug("voting block atx was timely",
+			log.FieldNamed("next_epoch", atxEpoch+1),
+			log.Time("next_epoch_start", nextEpochStart),
+			log.Time("atx_timestamp", atxTimestamp),
+			votingBlock.ID(),
+			votingBlock.ATXID,
+			log.Uint64("block_weight", blockWeight))
+		return blockWeight, nil
+	}
+	logger.With().Warning("voting block atx was untimely, zeroing block vote weight",
+		log.FieldNamed("next_epoch", atxEpoch+1),
+		log.Time("next_epoch_start", nextEpochStart),
+		log.Time("atx_timestamp", atxTimestamp),
+		votingBlock.ID(),
+		votingBlock.ATXID)
+	return 0, nil
+}
+
+func (t *turtle) voteWeightByID(ctx context.Context, votingBlockID, blockVotedOn types.BlockID) (uint64, error) {
+	block, err := t.bdp.GetBlock(votingBlockID)
+	if err != nil {
+		return 0, err
+	}
+	return t.voteWeight(ctx, block, blockVotedOn)
 }
 
 // Persist saves the current tortoise state to the database
 func (t *turtle) persist() error {
 	return t.bdp.Persist(mesh.TORTOISE, t)
 }
 
-// RecoverVerifyingTortoise retrieve latest saved tortoise from the database
+// RecoverVerifyingTortoise retrieves the latest saved tortoise from the database
 func RecoverVerifyingTortoise(mdb retriever) (interface{}, error) {
 	return mdb.Retrieve(mesh.TORTOISE, &turtle{})
 }
 
-func (t *turtle) processBlock(block *types.Block) error {
+func (t *turtle) processBlock(ctx context.Context, block *types.Block) error {
+	logger := t.logger.WithContext(ctx).WithFields(
```

Comment:
won't this add double key fields when logging line L563 or L564?

Reply:
yup, but I think it's a small price to pay for coherence in the logs throughout the code block